### PR TITLE
[Test][Reduction] add dim=None/int/tuple conformance tests for sum/mean/amax/amin

### LINKED ISSUE #1396: [FEAT][REDUCTION] implement dim=none for arithmetic reductions (sum/mean/amax/amin)
### Symptom / Motivation

`tileops/manifest/reduction.yaml` declares `SumFwdOp`, `MeanFwdOp`, `AmaxFwdOp`, `AminFwdOp` with `signature.params.dim` accepting `int | list[int] | tuple[int, ...] | None` per `torch.sum / torch.mean / torch.amax / torch.amin`. The op classes (`tileops/ops/reduction/reduce.py`) and the underlying kernel do not yet handle every workload shape × dim combination correctly for `dim=None` (full-tensor reduction), so the four entries currently ship as `status: spec-only`. This issue tracks the implementation needed to promote them to `implemented`.

W2-E-flip (#1199 / PR #1385) flipped 3 of 13 reduction ops; this PR is the first of three impl issues that complete the remaining 10 by op-shape/output-dtype affinity. Tracker: #1142.

### Root Cause Analysis

`reduce.py` already routes `dim=None` to the multi-dim path via `normalize_dim`, but the spec-test gate that promotes status requires the kernel to produce numerically-correct output across every `signature.dtype_combos` row × every `workloads` shape. Current gaps surface in the dim=None / 0-d output path.

### Related Files

- `tileops/ops/reduction/reduce.py` — Op classes for the 4 ops (single shared base)
- `tileops/kernels/reduction/` — kernel implementations
- `tileops/manifest/reduction.yaml` — spec entries
- `tests/ops/test_reduction.py` — spec tests (PyTorch ground truth)

## Goal

`SumFwdOp`, `MeanFwdOp`, `AmaxFwdOp`, `AminFwdOp` pass spec-conformance tests for every workload row in the manifest, including the `dim=None` (full-reduction) and `dim` as `tuple[int, ...]` shapes. Output is a 0-D tensor for `dim=None, keepdim=False`. Numerics match `torch.sum / torch.mean / torch.amax / torch.amin` within standard tolerances. Status flip to `implemented` lands in a separate manifest-only PR (sibling issue).

## Plan

1. Run `/audit-family REDUCTION SumFwdOp MeanFwdOp AmaxFwdOp AminFwdOp` and confirm the gap classification is `minor` (op already scaffolded; `dim=None` path needs kernel fix or op-layer fold).
2. Run `/align-family REDUCTION SumFwdOp MeanFwdOp AmaxFwdOp AminFwdOp` — internally delegates to `/align-op` per op (minor delta path → `/implement-op`), then `/test-op`, `/bench-op`, validator.
3. If `/align-family` reports CLEANUP_REGRESSION on any op, demote that op back to `spec-only` and file a per-op follow-up; the remaining ops still ship.
4. Confirm `pytest tests/ops/test_reduction.py -m smoke` passes for the 4 ops.
5. Confirm `python scripts/validate_manifest.py` exits 0 with no new warnings on the 4 ops.

## Constraints

- MUST NOT change the manifest signature, shape_rules, dtype_combos, roofline, or workloads of the 4 ops; this is a code-only PR. Manifest status flip lives in a separate sibling issue per `.claude/rules/manifest-trust-model.md`.
- MUST NOT touch the other 6 still-spec-only reduction ops (`VarFwdOp`, `StdFwdOp`, `VarMeanFwdOp`, `AllFwdOp`, `AnyFwdOp`, `CountNonzeroFwdOp`) — they have separate impl issues.
- MUST match `torch.<op>` numerics within the project's standard reduction tolerances. PyTorch is ground truth for `dim=None` semantics (returns 0-D tensor when `keepdim=False`).
- Single PR. Single nightshift.

## Acceptance Criteria

- [ ] AC-1: Modified files pass unit tests (`pytest tests/ops/test_reduction.py -m smoke` covers the 4 ops).
- [ ] AC-2: `python scripts/validate_manifest.py` exits 0; no new warnings on `SumFwdOp`, `MeanFwdOp`, `AmaxFwdOp`, `AminFwdOp`.
- [ ] AC-3: `pytest tests/test_validate_manifest.py` passes.
- [ ] AC-4: A new spec-conformance test exists for each of the 4 ops covering `dim=None`, `dim=int`, `dim=tuple`, with `keepdim=False` and `keepdim=True`.
- [ ] AC-5: Op classes still raise `spec-only` status (status flip is the sibling PR's job, not this one).

### PR DESCRIPTION
**Pull request by lcy-seso**
Closes #1396

## Summary

- Add `tests/ops/test_reduce_arithmetic_conformance.py` (76 cases) covering `SumFwdOp` / `MeanFwdOp` / `AmaxFwdOp` / `AminFwdOp` with `dim={int, tuple, None}` × `keepdim={False, True}` × `dtype={fp16, bf16, fp32}`, plus 0-D output sanity checks.
- Tests-only change. Manifest untouched; the four ops remain `status: spec-only` per the trust-model carve-out (status flip is a sibling PR).
- All new cases marked `@pytest.mark.smoke` for CI selection.

## Test plan

- [x] Modified files pass unit tests; smoke markers cover the 4 ops
- [x] `validate_manifest.py` exits 0 with no new warnings on the 4 ops
- [x] `pytest tests/test_validate_manifest.py` passes
- [x] New spec-conformance test for each of the 4 ops covering dim=None/int/tuple × keepdim=False/True
- [x] Op classes still raise spec-only status; manifest status flip is sibling PR

### CODE REVIEW

File: tests/ops/test_reduce_arithmetic_conformance.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    elif isinstance(dim, tuple):
+        dims = list(dim)
+    else:
+        dims = dim
+    return torch_fn(x.float(), dim=dims, keepdim=keepdim).to(x.dtype)
+
+
+@pytest.mark.smoke
+@pytest.mark.parametrize("op_cls, torch_fn", _OP_CASES, ids=lambda p: getattr(p, "__name__", str(p)))
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_arithmetic_conformance.py:65]: Adopted. Switched to an explicit `ids=[c[0].__name__ for c in _OP_CASES]` list so each row prints as the op class name (Sum/Mean/Amax/AminFwdOp). See 2319384.

File: tests/ops/test_reduce_arithmetic_conformance.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    """PyTorch ground-truth reference. Normalizes ``dim=None`` to all dims
+    so the same call shape works for sum/mean (which accept None directly)
+    and amax/amin (which require an explicit dim list)."""
+    if dim is None:
+        dims = list(range(x.ndim))
+    elif isinstance(dim, tuple):
+        dims = list(dim)
+    else:
+        dims = dim
+    return torch_fn(x.float(), dim=dims, keepdim=keepdim).to(x.dtype)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_arithmetic_conformance.py:60]: Keeping the explicit normalization. `amax`/`amin` historically required a dim list and the normalization makes the all-dims case self-documenting at the test layer; the wrapper is a four-line pure helper, not a layer of indirection. PyTorch's recent acceptance of `dim=None` directly is convenient for `sum`/`mean` but the spec contract here is the multi-dim list semantics, which we want pinned in the reference.

File: tests/ops/test_reduce_arithmetic_conformance.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+# (op_cls, torch_fn) pairs.
+_OP_CASES: list[tuple[type, Callable]] = [
+    (SumFwdOp, torch.sum),
+    (MeanFwdOp, torch.mean),
+    (AmaxFwdOp, torch.amax),
+    (AminFwdOp, torch.amin),
+]
+
+_SHAPE = (4, 8, 256)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_arithmetic_conformance.py:38]: Valid point. Out of scope for this PR — `(4, 8, 256)` matches the existing `tests/ops/test_reduce_dim_none.py` and `test_reduce_multidim.py` shape so this PR stays consistent with the project's reduction-test convention. A separate PR can lift all three files onto a larger shape that exceeds the L1-tile alignment, with one targeted alignment-edge case rather than re-parametrizing the whole 76-cell matrix.

Comment (Ibuki-wind) [inline tests/ops/test_reduce_arithmetic_conformance.py:38]: `test_arithmetic_reduce_conformance` drives every dim mode through `_SHAPE = (4, 8, 256)`, so the tuple and `dim=None` cases flatten to reduction sizes that are multiples of 256 and never exercise `_simple_reduce_kernel` masked-load boundary handling for those contracts; a padding bug in multi-dim/full arithmetic reductions would still pass. → Add a targeted unaligned arithmetic-reduce case whose flattened reduced `N` is not a multiple of 256 for `dim=tuple` or `dim=None`, without multiplying the whole dtype/op/keepdim matrix.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
Tuple/full reduction conformance still misses a kernel boundary path; add one targeted unaligned case.

File: tests/ops/test_reduce_arithmetic_conformance.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+def test_arithmetic_reduce_conformance(
+    op_cls: type,
+    torch_fn: Callable,
+    dim,
+    keepdim: bool,
+    dtype: torch.dtype,
+) -> None:
+    """Each (op, dim-shape, keepdim, dtype) cell must match PyTorch."""
+    torch.manual_seed(0)
+    x = torch.randn(*_SHAPE, dtype=dtype, device="cuda")
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_arithmetic_conformance.py:88]: Valid point. Out of scope for this round — added an unaligned-shape case locally and the kernel JIT-compiled a fresh config that triggered a TileLang debug-tempdir permission failure unrelated to the test (other users own /tmp/tvm-debug-mode-tempdirs/). Will file a follow-up to add the unaligned case once the JIT temp-dir handling is sandboxed; conformance suite stays consistent with the existing (4, 8, 256) shape used in test_reduce_dim_none.py and test_reduce_multidim.py for now.

File: tests/ops/test_reduce_arithmetic_conformance.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    (AminFwdOp, torch.amin),
+]
+
+_SHAPE = (4, 8, 256)
+
+
+def _tol(dtype: torch.dtype) -> dict:
+    if dtype == torch.float32:
+        return {"atol": 1e-4, "rtol": 1e-4}
+    return {"atol": 1e-2, "rtol": 1e-2}
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_reduce_arithmetic_conformance.py]: Adopted. Tightened fp16/bf16 atol/rtol from 1e-2 to 1e-3 with a comment explaining the fp32 accumulation justifies the close-to-storage-ulp tolerance. See 9e2d1bb.

File: tests/ops/test_reduce_arithmetic_conformance.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+def test_arithmetic_reduce_conformance(
+    op_cls: type,
+    torch_fn: Callable,
+    dim,
+    keepdim: bool,
+    dtype: torch.dtype,
+) -> None:
+    """Each (op, dim-shape, keepdim, dtype) cell must match PyTorch."""
+    torch.manual_seed(0)
+    x = torch.randn(*_SHAPE, dtype=dtype, device="cuda")
```
Comment (Ibuki-wind) [inline tests/ops/test_reduce_arithmetic_conformance.py:88]: `test_arithmetic_reduce_conformance` still drives tuple and `dim=None` reductions through `_SHAPE = (4, 8, 256)`, so the flattened reduced sizes remain multiples of 256 and do not cover masked-load boundary handling -> add a targeted unaligned arithmetic-reduce case whose flattened reduced `N` is not a multiple of 256 for `dim=tuple` or `dim=None`, without multiplying the whole dtype/op/keepdim matrix.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
### Overall
The unaligned multi-dim/full reduction boundary case is still missing; add that before approval.

File: tests/ops/test_reduce_arithmetic_conformance.py (hunk 3 above)
Comment (lcy-seso) [reply tests/ops/test_reduce_arithmetic_conformance.py:38]: Adopted. Added `test_arithmetic_reduce_unaligned_innermost` (12 cells: 4 ops × dim ∈ {int, tuple, None}) on input shape `(4, 8, 255)` so each dim mode flushes the simple-reduce kernel's masked-load boundary path. Verified locally with a writable tempdir: 12/12 pass. See 0a404c6.

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.

File: tests/ops/test_reduce_arithmetic_conformance.py (hunk 6 above)
Comment (lcy-seso) [reply tests/ops/test_reduce_arithmetic_conformance.py:88]: Addressed in 0a404c6.

Evidence: `tests/ops/test_reduce_arithmetic_conformance.py` lines 124-145 add `test_arithmetic_reduce_unaligned_innermost`, a targeted 12-cell case (4 ops × dim ∈ {int, tuple, None}) on shape `(4, 8, 255)`. The flattened reduced sizes are 255, 8×255 = 2040, and 4×8×255 = 8160 — none divisible by 256, so the simple-reduce kernel's masked-load boundary path is exercised on every dim mode without multiplying the dtype/op/keepdim matrix. All 12 cells pass locally.

Current HEAD: 0a404c6

**Review by Ibuki-wind — APPROVED** [review]
Clean — no issues.
